src/libpmemfile-posix/callbacks.c

```diff
@@ -0,0 +1,243 @@
+
+	/* Size of callbacks array. */
+	unsigned size;
+
+	/* Number of registered callbacks. */
+	unsigned used;
+};
+
+struct all_callbacks {
+	struct tx_callback_array forward;
+	struct tx_callback_array backward;
+};
+
+static os_tls_key_t callbacks_key;
+
+/*
+ * cb_get -- returns current per-thread callback configuration
+ */
+static struct all_callbacks *
+cb_get(void)
+{
+	struct all_callbacks *c = os_tls_get(callbacks_key);
+	if (!c) {
+		c = calloc(1, sizeof(struct all_callbacks) * MAX_TX_STAGE);
+		int ret = os_tls_set(callbacks_key, c);
+		if (ret) {
+			errno = ret;
+			FATAL("!os_tls_set");
+		}
+	}
+
+	return c;
+}
+
+/*
+ * cb_check -- (internal) check whether current state
+ * allows changing transaction callbacks
+ */
+static void
+cb_check(const char *func)
+{
+	if (pmemobj_tx_stage() == TX_STAGE_NONE)
+		FATAL("%s called outside of transaction", func);
+}
+
+/*
+ * cb_append -- (internal) appends callback to specified array
+ */
+static int
+cb_append(struct tx_callback_array *cb, cb_basic func, void *arg)
+{
+	if (cb->used == cb->size) {
+		unsigned count = cb->size * 2;
+		if (count == 0)
+			count = 4;
+
+		void *new_arr = realloc(cb->arr, count * sizeof(cb->arr[0]));
+		if (!new_arr) {
+			pmemfile_tx_abort(errno);
+			return -1;
+		}
+
+		cb->arr = new_arr;
+		cb->size = count;
+	}
+
+	unsigned num = cb->used++;
+	cb->arr[num].func = func;
+	cb->arr[num].arg = arg;
+
+	return 0;
+}
+
+/*
+ * cb_push_back -- registers transaction stage callback, to be called at the end
+ */
+int
+cb_push_back(enum pobj_tx_stage stage, cb_basic func, void *arg)
+{
+	LOG(15, NULL);
+
+	cb_check(__func__);
+	struct all_callbacks *cbs = cb_get();
+
+	return cb_append(&cbs[stage].forward, func, arg);
+}
+
+/*
+ * cb_push_front -- registers transaction stage callback, to be called at
+ * the beginning
+ */
+int
+cb_push_front(enum pobj_tx_stage stage, cb_basic func, void *arg)
+{
+	LOG(15, NULL);
+
+	cb_check(__func__);
+	struct all_callbacks *cbs = cb_get();
+
+	return cb_append(&cbs[stage].backward, func, arg);
+}
+
+/*
+ * cb_free -- frees tx callbacks
+ */
+static void
+cb_free(void *arg)
+{
+	LOG(15, NULL);
+	struct all_callbacks *callbacks = arg;
+
+	for (unsigned i = 0; i < MAX_TX_STAGE; ++i) {
+		free(callbacks[i].forward.arr);
+		callbacks[i].forward.arr = NULL;
+		callbacks[i].forward.size = 0;
+		callbacks[i].forward.used = 0;
+
+		free(callbacks[i].backward.arr);
+		callbacks[i].backward.arr = NULL;
+		callbacks[i].backward.size = 0;
+		callbacks[i].backward.used = 0;
+	}
+
+	free(callbacks);
+
+	int ret = os_tls_set(callbacks_key, NULL);
+	if (ret) {
+		errno = ret;
+		FATAL("!os_tls_set");
+	}
+}
+
+/*
+ * cb_init -- initializes callbacks subsystem
+ */
+void
+cb_init(void)
+{
+	int ret = os_tls_key_create(&callbacks_key, cb_free);
+	if (ret)
+		FATAL("!os_tls_key_create");
+}
+
+/*
+ * cb_fini -- cleans up state of tx callback module
+ */
+void
+cb_fini(void)
+{
+	struct all_callbacks *c = os_tls_get(callbacks_key);
+	if (c)
```

Comment:
cstyle is executed for all c files when DEVELOPER_MODE=1. Cstyle passes for this file. This is the same style as used in NVML.